### PR TITLE
Fix System.Runtime tests on non EN cultures

### LINKED ISSUE #8753: String formatting tests fail when run on some cultures
When I run `msbuild /t:BuildAndTest System.Runtime.Tests.csproj`, I get many failures in string formatting tests (including `StringBuilder` and `.Parse()`). The cause seems to be culture: the tests appear to run in my system culture, which uses `,` as decimal separator, while the tests expect `.`. 

[Tests output](https://gist.github.com/svick/383824511c8778d57e959355082e35fd)

I expect the fix is to force the thread to invariant culture for those tests.

### PR DESCRIPTION
**Pull request by hughbe**
- Fix DateTimeOffset tests to create dates from DateTimeKind.Utc not
DateTimeKind.Unspecified (I think this is because we previously adjusted for time zones causing a negative number)
- Fix tests for Parse for Decimal, Double and Single to obey the
IFormatProvider parameter
- Add a helper to convert a string containing an English decimal ('.')
to the current culture

I didn't want to turn these tests into using the Invariant culture, as the tests by nature should have different results on different cultures and we want to make sure that everything works.

I've ran tests by setting my culture to the Czech Repuplic (cs-CZ) as this was the culture in the issue reported. Originally there were failures, now there are none. 

Fixes #8753

/cc @tarekgh @stephentoub @svick

### CODE REVIEW

**Comment by ellismg** [comment]
> I didn't want to turn these tests into using the Invariant culture, as the tests by nature should have different results on different cultures and we want to make sure that everything works.

Just my two cents, but I disagree with this approach.  If there are concerns we have around using invariant specifically (e.g we want a culture with different separators or groupings our whatever), we should create a synthetic culture with different data, but I do believe we should be using fixed data for tests.  I think trying to write tests in a style where there robust against changes in culture data is not a good idea.  I'd be interested in what @tarekgh thinks.

**Comment by hughbe** [comment]
My concern is that creating a synthetic test data means that we never test the simple `int.Parse(string)` method and this never verify what happens for the most common use case. I'm not too bothered though

**Comment by hughbe** [comment]
Test Innerloop CentOS7.1 Debug Build and Test 
Test Innerloop Windows_NT Debug Build and Test

**Comment by ellismg** [comment]
> My concern is that creating a synthetic test data means that we never test the simple int.Parse(string) method and this never verify what happens for the most common use case. 

My preferred option for testing these sorts of methods would be to set an explicit known CurrentCulture (or CurrentUICulture if the method pulls that instead) before calling the method and then reset it when the test is done.

> I'm not sure what you mean by changes in culture data: this already responds to culture data, so any changes in the culture data will be reflected by the LocalizeDecimalString method.

I do believe what you have here will work for the cases your throwing at it.  If we want to use ambient culture data when testing, I think what you've done is a good way to solve the problem we're facing.  My point was more around the fact that we'll be chasing stuff all over in the limit if we try to convert expected data from one culture to another in order to verify things.

**Comment by hughbe** [comment]
> My preferred option for testing these sorts of methods would be to set an explicit known CurrentCulture (or CurrentUICulture if the method pulls that instead) before calling the method and then reset it when the test is done.

Ah, I see. I honestly don't mind - does that require all that RemoteInvoke code so that culture changes don't affect other tests?

**Comment by ellismg** [comment]
> Ah, I see. I honestly don't mind - does that require all that RemoteInvoke code so that culture changes don't affect other tests?

It shouldn't.  You can just do the following:

```csharp

CultureInfo oldCulture = CultureInfo.CurrentCulture;
try
{
    CultureInfo.CurrentCulture = new CultureInfo("en-US");
    ... stuff to test ...
}
finally
{
    CultureInfo.CurrentCulture = oldCulture;
}
```

**Comment by hughbe** [comment]
Oh right - I'll wait for @tarekgh to respond to whether its fine as is or should be changes before I go changing code up

**Comment by bbowyersmyth** [comment]
There is the `PerformActionWithCulture` helper in StringTests.cs if you like

**Comment by tarekgh** [comment]
For the point of settings current culture, we already have the way to isolate the test to run in its own process as we do in System.Globalization tests. if you are going to set the current culture, it'll be nice to follow that mechanism. 

regarding the point of should we test with fixed data or testing whatever culture is set, my opinion here is in the middle. I mean it depends on what is we are testing. for example, if we need to test the parsing, we can just format a number and then testing parsing (regardless of current culture). by doing that we shouldn't care what is culture data is really used.
if we testing specific culture data, I would agree with Matt that we should have fixed data we are testing against.

to summarize regarding the specified tests we are changing here, for testing formatting, we should test against fixed data (as Matt suggested). Also I suggest if you'll set the current culture, we move such tests to System.Globalization test as it is already protected against setting current culture.
For Parse testing, we should format the string and then parse it without caring about what is the culture is used and what is the decimal separator or any other culture data.

**Comment by tarekgh** [comment]
@hughbe just checking if you are going to update the changes here? Thanks.

**Comment by hughbe** [comment]
Yeah sorry I was preoccupied. Just so I don't get confused, you're suggesting I do the following:
- For the ToString and StringBuilder tests: change the current culture while running the tests
- For the Parse tests: keep as is

Cheers

**Comment by tarekgh** [comment]
@hughbe 

> For the ToString and StringBuilder tests: change the current culture while running the tests

Yes but I would suggest to move such tests to System.Globalization as it uses RemoteExecutorConsoleApp for isolation. also I am seeing this related to globalization area too. 


> For the Parse tests: keep as is

For parsing we need to get the expected value by formatting the number. think about it as round tripping test.  so we don't need LocalizeDecimalString at all. 

please let me know if there is anything unclear here. and thanks for your effort here.

**Comment by tarekgh** [comment]
@hughbe I thought more about the formatting tests so I am ok to keep the tests here and not moving it to globalization tests. just apply Matt's recommendation to set the current culture.

**Comment by hughbe** [comment]
Brilliant: I've done this. I ran the tests again locally on my PC set the Czech and UK cultures and they work.
@bbowyersmyth thanks - I moved this into the helpers file

**Comment by tarekgh** [comment]
LGTM. thanks a lot @hughbe

**Comment by tarekgh** [comment]
test  Innerloop Windows_NT Debug Build and Tes please

**Comment by tarekgh** [comment]
test Innerloop Windows_NT Debug Build and Tes please
